Firestore/core/src/firebase/firestore/api/snapshot_metadata.cc

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/api/snapshot_metadata.h"
+
+#include "Firestore/core/src/firebase/firestore/util/hashing.h"
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+bool operator==(const SnapshotMetadata& lhs, const SnapshotMetadata& rhs) {
+  return lhs.pending_writes_ == rhs.pending_writes_ &&
+         lhs.from_cache_ == rhs.from_cache_;
+}
+
+size_t SnapshotMetadata::Hash() const {
+  return util::Hash(pending_writes_, from_cache_);
```

Comment:
For my own curiosity (not asking you to change anything), why are pending_writes() and from_cache() in the .h but Hash() is in the .mm?

https://google.github.io/styleguide/cppguide.html#Inline_Functions suggests that "< 10 lines of code" is a good criteria for inlining.  But maybe we're following a different convention or that doesn't apply for some reason.

Reply:
The issue is that `util::Hash` looks like compact code but the template magic required to make it work is nontrivial. Including `hashing.h` in the .cc file isolates the rest of the code from also including and parsing hashing.h just to see SnapshotMetadata.

Anyway, my general rule is that if inlining a function requires the addition of a header to another header it may be worth avoiding the inline.

For contrast `pending_writes()` and `from_cache()` are "obviously" just returning a simple field so there's no extra cost to allowing the compiler to inline a field access.